API/Repositories/Implementations/HistoryRepositoryBase.cs

```diff
@@ -0,0 +1,70 @@
+﻿using API.Entities.Interfaces;
+using API.Enums;
+using API.Repositories.Interfaces;
+using AutoMapper;
+using Microsoft.EntityFrameworkCore;
+
+namespace API.Repositories.Implementations;
+
+public class HistoryRepositoryBase<TEntity, THistory>
+    : RepositoryBase<TEntity>,
+        IHistoryRepository<TEntity, THistory>
+    where THistory : class, IHistoryEntity
+    where TEntity : class, IEntityBase
+{
+    private readonly OtrContext _context;
+    private readonly IMapper _mapper;
+
+    protected HistoryRepositoryBase(OtrContext context, IMapper mapper)
+        : base(context)
+    {
+        _mapper = mapper;
+        _context = context;
+    }
+
+    public override async Task<int> UpdateAsync(TEntity entity)
+    {
+        await CreateHistoryAsync(entity.Id, HistoryActionType.Update);
+        entity.Updated = DateTime.UtcNow;
+        return await base.UpdateAsync(entity);
+    }
+
+    public override async Task<int?> DeleteAsync(int id)
+    {
+        await CreateHistoryAsync(id, HistoryActionType.Delete);
+        return await base.DeleteAsync(id);
+    }
+
+    public async Task<int> UpdateAsync(TEntity entity, int modifierId)
+    {
+        await CreateHistoryAsync(entity.Id, HistoryActionType.Update, modifierId);
+        entity.Updated = DateTime.UtcNow;
+        return await base.UpdateAsync(entity);
+    }
+
+    public async Task<int?> DeleteAsync(int id, int modifierId)
+    {
+        await CreateHistoryAsync(id, HistoryActionType.Delete, modifierId);
+        return await base.DeleteAsync(id);
+    }
+
+    private async Task<TEntity?> GetNoTrackingAsync(int id) =>
+        await _context.Set<TEntity>().AsNoTracking().FirstOrDefaultAsync(x => x.Id == id);
+
+    private async Task<THistory?> CreateHistoryAsync(int id, HistoryActionType action, int? modifierId = null)
+    {
+        TEntity? origEntity = await GetNoTrackingAsync(id);
+        if (origEntity == null)
+            return null;
```

Comment:
I was thinking this should probably throw an error since realistically if the entity doesn't exist when we get to this point then something probably went wrong, or something is making changes to context outside of the standard CRUD. Which is, and I quote "sussy wussy"

Reply:
I feel like that should be the responsibility of the caller to handle. Returning null *is* the unexpected behavior here.